### PR TITLE
add clipping to square-distance calculation

### PR DESCRIPTION
**Pull request by rsokl**

### CODE REVIEW

File: Python/Module3_IntroducingNumpy/Broadcasting.md [hunk 1]
```diff
@@ -620,34 +638,106 @@ Having accounted for all three terms, we can finally compute the Euclidean dista
...
 ```python
 # computing all the distances
->>> dists = np.sqrt(x_y_sqrd + x_y_prod)
+>>> dists = np.sqrt(x_y_sqrd - x_y_prod)
 >>> dists.shape
 (5, 6)
 ```
-In total, we have successfully used vectorization to compute the all pairs of distances, while only requiring an array of shape-$(M, N)$ to do so! This is the memory-efficient, vectorized form - the stuff that dreams are made of. Let's write the function that performs this computation in full.  
+
+#### A Subtle Issue with Floating Point Precision
```
Comment (davidmascharka) [inline Python/Module3_IntroducingNumpy/Broadcasting.md:646]: nit:
```suggestion
#### A Subtle Issue with Floating-point Precision
```
or
```suggestion
#### A Subtle Issue with Floating-Point Precision
```

File: Python/Module3_IntroducingNumpy/Broadcasting.md [hunk 2]
```diff
@@ -620,34 +638,106 @@ Having accounted for all three terms, we can finally compute the Euclidean dista
...
+
+There is one more important and very subtle detail that we have to deal with.
+In terms of pure mathematics, `x_y_sqrd - x_y_prod` must be a strictly non-negative value (i.e. its smallest possible value is $0$), since it is equivalent to
+
+\begin{equation}
+\sum_{i=0}^{D-1}{(x_{i} - y_{i})^2}
+\end{equation}
+
+That being said, are working with floating point numbers, which do not always behave exactly like rational numbers when we do arithmetic on them.
```
Comment (davidmascharka) [inline Python/Module3_IntroducingNumpy/Broadcasting.md:655]: nit:
```suggestion
That being said, are working with floating-point numbers, which do not always behave exactly like rational numbers when we do arithmetic with them.
```

File: Python/Module3_IntroducingNumpy/Broadcasting.md [hunk 3]
```diff
@@ -620,34 +638,106 @@ Having accounted for all three terms, we can finally compute the Euclidean dista
...
+Indeed, we saw earlier that the [quirks of floating point arithmetic](https://www.pythonlikeyoumeanit.com/Module2_EssentialsOfPython/Basic_Objects.html#Understanding-Numerical-Precision) can lead to surprising results.
```
Comment (davidmascharka) [inline Python/Module3_IntroducingNumpy/Broadcasting.md:656]: nit:
```suggestion
Indeed, we saw earlier that the [quirks of floating-point arithmetic](https://www.pythonlikeyoumeanit.com/Module2_EssentialsOfPython/Basic_Objects.html#Understanding-Numerical-Precision) can lead to surprising results. 
```

File: Python/Module3_IntroducingNumpy/Broadcasting.md [hunk 4]
```diff
@@ -620,34 +638,106 @@ Having accounted for all three terms, we can finally compute the Euclidean dista
...
+
+Let's see this in action.
+We'll take the following shape-(2, 3) array
 
 ```python
-def pairwise_dists(x, y):
-    """ Computing pairwise distances using memory-efficient 
-        vectorization.
+# A carefully-selected input that will trigger surprising
+# floating point precision issues in our distance calculation
```
Comment (davidmascharka) [inline Python/Module3_IntroducingNumpy/Broadcasting.md:664]: nit:
```suggestion
# floating-point precision issues in our distance calculation
```

File: Python/Module3_IntroducingNumpy/Broadcasting.md [hunk 5]
```diff
@@ -620,34 +638,106 @@ Having accounted for all three terms, we can finally compute the Euclidean dista
...
+>>> x = np.array([[4.700867387959219, 4.700867387959219, 4.700867387959219],
+...               [4.700867387959219, 4.700867387959219, 4.700867387959219]])
+```
 
-        Parameters
-        ----------
-        x : numpy.ndarray, shape=(M, D)
-        y : numpy.ndarray, shape=(N, D)
+And compute the square distance between all combination of rows `x`; the result should simply a shape-(2, 2) array of zeros.
```
Comment (davidmascharka) [inline Python/Module3_IntroducingNumpy/Broadcasting.md:669]:
```suggestion
And compute the square distance between all combinations of rows `x`; the result should simply be a shape-(2, 2) array of zeros.
```

File: Python/Module3_IntroducingNumpy/Broadcasting.md [hunk 6]
```diff
@@ -620,34 +638,106 @@ Having accounted for all three terms, we can finally compute the Euclidean dista
...
+# will instead be (*very* small) negative numbers!
+>>> sqr_dists = -2 * np.matmul(x, x.T)  # `x_x_prod`
+>>> sqr_dists +=  np.sum(x**2, axis=1)[:, np.newaxis]
+>>> sqr_dists += np.sum(x**2, axis=1)
+>>> sqr_dists
+array([[-2.842170943040401e-14, -2.842170943040401e-14],
+       [-2.842170943040401e-14, -2.842170943040401e-14]])
 ```
 
+These values are *very* close to being zero, so it is not as if the magnitude of our result is wildly off, but the critical issue here is that the quirks of floating point arithmetic produced (very small) negative numbers.
```
Comment (davidmascharka) [inline Python/Module3_IntroducingNumpy/Broadcasting.md:682]: nit:
```suggestion
These values are *very* close to being zero, so it is not as if the magnitude of our result is wildly off, but the critical issue here is that the quirks of floating-point arithmetic produced (very small) negative numbers.
```

File: Python/Module3_IntroducingNumpy/Broadcasting.md [hunk 7]
```diff
@@ -620,34 +638,106 @@ Having accounted for all three terms, we can finally compute the Euclidean dista
...
+>>> np.sqrt(np.clip(sqr_dists, a_min=0, a_max=None))
+array([[0., 0.],
+       [0., 0.]])
+```
+<!-- #endregion -->
+
+<!-- #region -->
+#### The Final Answer, At Last!
+
+In total, we have successfully used vectorization to compute the all pairs of distances, while only requiring an array of shape-$(M, N)$ to do so! This is the memory-efficient, vectorized form – the stuff that dreams are made of.
```
Comment (davidmascharka) [inline Python/Module3_IntroducingNumpy/Broadcasting.md:708]:
```suggestion
In total, we have successfully used vectorization to compute all the pairs of distances, while only requiring an array of shape-$(M, N)$ to do so! This is the memory-efficient, vectorized form – the stuff that dreams are made of.
```

File: Python/Module3_IntroducingNumpy/Broadcasting.md [hunk 8]
```diff
@@ -620,34 +638,106 @@ Having accounted for all three terms, we can finally compute the Euclidean dista
...
+    Parameters
+    ----------
+    x : numpy.ndarray, shape=(M, D)
+    y : numpy.ndarray, shape=(N, D)
+
+    Returns
+    -------
+    numpy.ndarray, shape=(M, N)
+        The Euclidean distance between each pair of
+        rows between `x` and `y`."""
```
Comment (davidmascharka) [inline Python/Module3_IntroducingNumpy/Broadcasting.md:725]: do you want to leave the end-quote on the same line or a newline? not sure what the consistent thing is off the top of my head

Comment (rsokl) [reply Python/Module3_IntroducingNumpy/Broadcasting.md:725]: I think same line is what `black` does. So I'll go with it.

**Comment by rsokl** [comment]
Thank you @davidmascharka ! It's been a while since we did a PR here :)

**Pull request merged**